### PR TITLE
Unify Lucene document creation

### PR DESCRIPTION
**Pull request by joelverhagen**
1. Renamed `EditedDate` to `LastEditedDate` (since metadata has `lastEdited`). This mirrors `published` to `PublishedDate`.
2. Renamed `DisplayName` to `SortableTitle` which better suits the use.
3. Renamed `flattenedAuthors` to `authors` because it is always flattened (unlike `Owner`).
4. Add additional validation to document fields that are parsed during result materialization.
   - `Listed` must be `Boolean` parsable.
   - `RequiresLicenseAcceptance` must be `Boolean` parsable.
   - `PackageSize` must be `Int32` parsable.
5. Remove `LicenseNames` and `LicenseReportUrl` since they are never used by the gallery.
6. Default a missing `title` metadata field to the `id` value (which is required).
7. Add unit tests.
8. Unify Lucene document creation in "DB to Lucene" and "Catalog to Lucene" scenarios.

### CODE REVIEW

File: src/Ng/SearchIndexFromCatalogCollector.cs [hunk 1]
```diff
@@ -269,167 +255,6 @@ static float DetermineLanguageBoost(string id, string language)
...
-            {
-                NormalizeId(catalogEntry);
-
-                return CreateLuceneDocument_ApiApp(catalogEntry, packageUrl, baseAddress);
-            }
-
-            return CreateLuceneDocument_NuGet(catalogEntry, packageUrl);
-        }
-
-        static Document CreateLuceneDocument_ApiApp(JObject package, string packageUrl, string baseAddress)
```
Comment (joelverhagen) [inline src/Ng/SearchIndexFromCatalogCollector.cs]: I am removing API app-specific code here. Is this okay? @maartenba

File: src/NuGet.Services.BasicSearch/Web.config [hunk 2]
```diff
@@ -13,12 +13,12 @@
     <customErrors mode="Off" />
   </system.web>
   <appSettings>
+    <!--
```
Comment (joelverhagen) [inline src/NuGet.Services.BasicSearch/Web.config]: This was breaking DEV.

File: src/Ng/SearchIndexFromCatalogCollector.cs (hunk 1 above)
Comment (maartenba) [inline src/Ng/SearchIndexFromCatalogCollector.cs]: By all means. Lots more of that stuff in the project (check with @johnataylor)

File: src/Ng/SearchIndexFromCatalogCollector.cs [hunk 3]
```diff
@@ -53,7 +53,7 @@ public SearchIndexFromCatalogCollector(Uri index, Lucene.Net.Store.Directory dir
 
                 indexWriter.ExpungeDeletes();
 
-                indexWriter.Commit(CreateCommitMetadata(commitTimeStamp));
+                indexWriter.Commit(DocumentCreator.CreateCommitMetadata(commitTimeStamp, "from catalog", Guid.NewGuid().ToString()));
```
Comment (joyhui) [inline src/Ng/SearchIndexFromCatalogCollector.cs]: Please make sure you have test to confirm the component who reading this metadata compatible with this change.

File: src/Ng/SearchIndexFromCatalogCollector.cs [hunk 4]
```diff
@@ -269,167 +255,6 @@ static float DetermineLanguageBoost(string id, string language)
...
-
-            Add(doc, "OriginalVersion", (string)package["verbatimVersion"], Field.Store.YES, Field.Index.NOT_ANALYZED, Field.TermVector.NO);
-            Add(doc, "OriginalCreated", (string)package["created"], Field.Store.YES, Field.Index.NOT_ANALYZED, Field.TermVector.NO);
-            Add(doc, "OriginalLastEdited", (string)package["lastEdited"], Field.Store.YES, Field.Index.NOT_ANALYZED, Field.TermVector.NO);
-            Add(doc, "OriginalPublished", (string)package["published"], Field.Store.YES, Field.Index.NOT_ANALYZED, Field.TermVector.NO);
-
-            Add(doc, "LicenseNames", (string)package["licenseNames"], Field.Store.YES, Field.Index.NOT_ANALYZED, Field.TermVector.NO);
-            Add(doc, "LicenseReportUrl", (string)package["licenseReportUrl"], Field.Store.YES, Field.Index.NOT_ANALYZED, Field.TermVector.NO);
-
-            //TODO: add dependency summary
```
Comment (joelverhagen) [inline src/Ng/SearchIndexFromCatalogCollector.cs]: We need the dependencies and supported frameworks for frameworks for V2 compatibility, correct?

File: src/Ng/SearchIndexFromCatalogCollector.cs (hunk 3 above)
Comment (joelverhagen) [inline src/Ng/SearchIndexFromCatalogCollector.cs]: Good point, I missed the `commit-time-stamp` property which is used in one place of our code.

**Comment by joyhui** [comment]
:shipit:

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                AddString("releaseNotes");
+                AddString("copyright");
+                AddString("language");
+                AddString("licenseUrl");
+                AddString("packageHash");
+                AddString("packageHashAlgorithm");
+                AddString("packageSize");
+                AddString("requiresLicenseAcceptance");
+
+                // TODO: add flattenedDependencies and supportedFrameworks
```
Comment (maartenba) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: Fix todo before :shipit:

File: src/NuGet.Indexing/Extraction/DocumentCreator.cs [hunk 6]
```diff
@@ -26,28 +27,36 @@ public static IndexWriter CreateIndexWriter(Lucene.Net.Store.Directory directory
             return indexWriter;
         }
 
-        public static IDictionary<string, string> CreateCommitMetadata(string description, string trace)
+        public static IDictionary<string, string> CreateCommitMetadata(DateTime commitTimeStamp, string description, string trace)
         {
             IDictionary<string, string> commitMetadata = new Dictionary<string, string>();
-            commitMetadata.Add("commitTimeStamp", DateTime.UtcNow.ToString("O"));
+            commitMetadata.Add("commitTimeStamp", commitTimeStamp.ToString("O"));
+            commitMetadata.Add("commit-time-stamp", commitTimeStamp.ToString("O"));
```
Comment (maartenba) [inline src/NuGet.Indexing/Extraction/DocumentCreator.cs]: Why both properties? Can we standardize on one?

File: src/NuGet.Indexing/Extraction/DocumentCreator.cs [hunk 7]
```diff
@@ -128,110 +119,101 @@ private static void AddVersion(Document document, IDictionary<string, string> pa
                     }
                     else
                     {
-                        errors.Add("Unable to parse 'originalVersion' property.");
+                        errors.Add("Unable to parse 'originalVersion' as NuGetVersion.");
```
Comment (maartenba) [inline src/NuGet.Indexing/Extraction/DocumentCreator.cs]: Are errors logged somewhere? What happens to the errors variable if we crash out, do we still get them in logs?

**Comment by maartenba** [comment]
Few remarks, other than those looks great!
:

File: src/Ng/SearchIndexFromCatalogCollector.cs (hunk 4 above)
Comment (joelverhagen) [inline src/Ng/SearchIndexFromCatalogCollector.cs]: Fixed.

File: src/NuGet.Indexing/Extraction/DocumentCreator.cs (hunk 6 above)
Comment (joelverhagen) [inline src/NuGet.Indexing/Extraction/DocumentCreator.cs]: I've done a lot of investigation right now and it seems like we have [some diagnostic/monitoring](https://github.com/search?l=&q=commit-time-stamp+user%3Anuget+language%3AC%23&ref=advsearch&type=Code&utf8=%E2%9C%93) tools that depend on this.

**Pull request merged**